### PR TITLE
refactor: separate payload processing from payloads WPB-4965

### PR DESCRIPTION
**Pull request by johnxnguyen**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-4965" title="WPB-4965" target="_blank">TaskWPB-4965</a>  [iOS] separate payload processing logic from payloads
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

A lot of code that handles the processing of response payloads are defined as extensions of the payloads themselves. This makes it hard to test and mock the logic because it is tightly coupled with the data it works with. In this PR we separate the logic from the data into separate "payload processors". 

#### Test Coverage 

- Tests have been moved to the corresponding payload processor tests.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-request-strategy/Sources/Payloads/Processing/ConversationEventPayloadProcessor.swift [hunk 1]
```diff
@@ -0,0 +1,777 @@
...
+            conversation = existingConversation
+        } else {
+            conversation = ZMConversation.fetchOrCreate(with: conversationID, domain: payload.qualifiedID?.domain, in: context)
+            otherUser.connection?.conversation = conversation
+        }
+
+        conversation.remoteIdentifier = conversationID
+        conversation.domain = BackendInfo.isFederationEnabled ? payload.qualifiedID?.domain : nil
+
+        // TODO: use conversation type from the backend once it returns the correct value
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Payloads/Processing/ConversationEventPayloadProcessor.swift]: is it still relevant ? if yes, could we link it to a jira ticket or solve it already

File: wire-ios-request-strategy/Sources/Payloads/Processing/ConversationEventPayloadProcessor.swift [hunk 2]
```diff
@@ -0,0 +1,777 @@
...
+
+        conversation.updateMembers(otherMembers, selfUserRole: selfUserRole)
+    }
+
+    func updateConversationTimestamps(
+        for conversation: ZMConversation,
+        serverTimestamp: Date
+    ) {
+        // If the lastModifiedDate is non-nil, e.g. restore from backup, do not update the lastModifiedDate
+        if conversation.lastModifiedDate == nil { // TODO jacob review this logic
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Payloads/Processing/ConversationEventPayloadProcessor.swift]: same I guess we can remove this TODO

**Review by netbe — APPROVED** [review]
LGTM;) just clean up some ported TODOs that are maybe obsolete

File: wire-ios-request-strategy/Sources/Payloads/Processing/ConnectionPayloadProcessorTests.swift [hunk 3]
```diff
@@ -18,16 +19,32 @@
...
 
-class PayloadProcessing_ConnectionTests: MessagingTestBase {
+final class ConnectionPayloadProcessorTests: MessagingTestBase {
+
+    var sut: ConnectionPayloadProcessor!
+
+    override func setUp() {
+        super.setUp()
+        sut = ConnectionPayloadProcessor()
+        BackendInfo.storage = .random()!
```
Comment (David-Henner) [inline wire-ios-request-strategy/Sources/Payloads/Processing/ConnectionPayloadProcessorTests.swift:29]: does this need to be reset in teardown?

File: wire-ios-request-strategy/Sources/Payloads/Processing/ConversationEventPayloadProcessor.swift [hunk 4]
```diff
@@ -0,0 +1,777 @@
...
+final class ConversationEventPayloadProcessor {
+
+    enum Source {
+        case slowSync
+        case eventStream
+    }
+
+    // MARK: - Conversation creation
+
+    func updateOrCreateConverations(
```
Comment (David-Henner) [inline wire-ios-request-strategy/Sources/Payloads/Processing/ConversationEventPayloadProcessor.swift]:
```suggestion
    func updateOrCreateConversations(
```

File: wire-ios-request-strategy/Sources/Payloads/Processing/ConversationEventPayloadProcessor.swift [hunk 5]
```diff
@@ -0,0 +1,777 @@
...
+        for payload in payload.conversations {
+            updateOrCreateConversation(
+                from: payload,
+                source: .slowSync,
+                in: context
+            )
+        }
+    }
+
+    func updateOrCreateConverations(
```
Comment (David-Henner) [inline wire-ios-request-strategy/Sources/Payloads/Processing/ConversationEventPayloadProcessor.swift:43]:
```suggestion
    func updateOrCreateConversations(
```

File: wire-ios-request-strategy/Sources/Payloads/Processing/ConversationEventPayloadProcessor.swift (hunk 1 above)
Comment (johnxnguyen) [reply wire-ios-request-strategy/Sources/Payloads/Processing/ConversationEventPayloadProcessor.swift]: Yeah it's still a problem. I'll see if backend has a ticket for this.

Comment (johnxnguyen) [reply wire-ios-request-strategy/Sources/Payloads/Processing/ConversationEventPayloadProcessor.swift]: Actually this is now the way to do it, according to https://wearezeta.atlassian.net/browse/WPB-3560. So I'll remove the TODO.

File: wire-ios-request-strategy/Sources/Payloads/Processing/ConnectionPayloadProcessorTests.swift (hunk 3 above)
Comment (johnxnguyen) [reply wire-ios-request-strategy/Sources/Payloads/Processing/ConnectionPayloadProcessorTests.swift:29]: fixed

File: wire-ios-request-strategy/Sources/Payloads/Processing/ConversationEventPayloadProcessor.swift (hunk 4 above)
Comment (johnxnguyen) [reply wire-ios-request-strategy/Sources/Payloads/Processing/ConversationEventPayloadProcessor.swift]: fixed

File: wire-ios-request-strategy/Sources/Payloads/Processing/ConversationEventPayloadProcessor.swift (hunk 5 above)
Comment (johnxnguyen) [reply wire-ios-request-strategy/Sources/Payloads/Processing/ConversationEventPayloadProcessor.swift:43]: fixed

**Review by David-Henner — APPROVED** [review]

**Pull request merged**